File: src/main/java/me/dyzjct/kura/module/modules/misc/AntiPiston.java
package me.dyzjct.kura.module.modules.misc;

import me.dyzjct.kura.module.Category;
import me.dyzjct.kura.module.Module;
import me.dyzjct.kura.setting.Setting;
import me.dyzjct.kura.utils.NTMiku.BlockUtil;
import me.dyzjct.kura.utils.inventory.InventoryUtil;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityEnderCrystal;
import net.minecraft.init.Blocks;
import net.minecraft.network.play.client.CPacketAnimation;
import net.minecraft.network.play.client.CPacketHeldItemChange;
import net.minecraft.network.play.client.CPacketUseEntity;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;

import java.util.Comparator;
import java.util.stream.Collectors;

@Module.Info(name = "AntiPiston", category = Category.MISC)
public class AntiPiston extends Module {
    private final Setting<Boolean> rotate = bsetting("Rotate", false);
    private int obsidian = -1;

    public static void breakcrystal() {
        if (fullNullCheck()) return;
        for (Entity crystal : AntiPiston.mc.world.loadedEntityList.stream().filter(e -> e instanceof EntityEnderCrystal && !e.isDead).sorted(Comparator.comparing(e -> Float.valueOf(AntiPiston.mc.player.getDistance(e)))).collect(Collectors.toList())) {
            if (!(crystal instanceof EntityEnderCrystal) || !(AntiPiston.mc.player.getDistance(crystal) <= 4.0f))
                continue;
            AntiPiston.mc.player.connection.sendPacket(new CPacketUseEntity(crystal));
            AntiPiston.mc.player.connection.sendPacket(new CPacketAnimation(EnumHand.OFF_HAND));
        }
    }

    @Override
    public void onEnable() {
        if (fullNullCheck()) return;
        AntiPiston.breakcrystal();
    }

    @Override
    public void onUpdate() {
        if (fullNullCheck()) return;
        if (AntiPiston.mc.player == null || AntiPiston.mc.world == null) {
            return;
        }
        this.obsidian = InventoryUtil.findHotbarBlock(Blocks.OBSIDIAN);
        if (this.obsidian == -1) {
            return;
        }
        BlockPos pos = new BlockPos(AntiPiston.mc.player.posX, AntiPiston.mc.player.posY, AntiPiston.mc.player.posZ);
        if (pos == null) {
            return;
        }
        if (this.getBlock(pos.add(1, 1, 0)).getBlock() == Blocks.PISTON) {
            if (this.getBlock(pos.add(-1,1,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(-1, 1, 0));
            }
            else if (this.getBlock(pos.add(-1,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(-1, 2, 0));
            }
            else if (this.getBlock(pos.add(0,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0,2,0));
            }
            mc.playerController.onPlayerDamageBlock(pos.add(1, 1, 0), BlockUtil.getRayTraceFacing(pos.add(1, 1, 0)));
        }
        if (this.getBlock(pos.add(-1, 1, 0)).getBlock() == Blocks.PISTON) {
            if (this.getBlock(pos.add(1,1,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(1, 1, 0));
            }
            else if (this.getBlock(pos.add(1,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(1, 2, 0));
            }
            else if (this.getBlock(pos.add(0,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0,2,0));
            }
            mc.playerController.onPlayerDamageBlock(pos.add(-1, 1, 0), BlockUtil.getRayTraceFacing(pos.add(-1, 1, 0)));
        }
        if (this.getBlock(pos.add(0, 1, 1)).getBlock() == Blocks.PISTON) {
            if (this.getBlock(pos.add(0,1,-1)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0, 1, -1));
            }
            else if (this.getBlock(pos.add(0,2,-1)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0, 2, -1));
            }
            else if (this.getBlock(pos.add(0,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0,2,0));
            }
            mc.playerController.onPlayerDamageBlock(pos.add(0, 1, 1), BlockUtil.getRayTraceFacing(pos.add(0, 1, 1)));
        }
        if (this.getBlock(pos.add(0, 1, -1)).getBlock() == Blocks.PISTON) {
            if (this.getBlock(pos.add(0,1,1)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0, 1, 1));
            }
            else if (this.getBlock(pos.add(0,2,1)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0, 2, 1));
            }
            else if (this.getBlock(pos.add(0,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0,2,0));
            }
            mc.playerController.onPlayerDamageBlock(pos.add(0, 1, -1), BlockUtil.getRayTraceFacing(pos.add(0, 1, -1)));
        }
//        STICKY_PISTON
        if (this.getBlock(pos.add(1, 1, 0)).getBlock() == Blocks.STICKY_PISTON) {
            if (this.getBlock(pos.add(-1,1,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(-1, 1, 0));
            }
            else if (this.getBlock(pos.add(-1,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(-1, 2, 0));
            }
            else if (this.getBlock(pos.add(0,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0,2,0));
            }
            mc.playerController.onPlayerDamageBlock(pos.add(1, 1, 0), BlockUtil.getRayTraceFacing(pos.add(1, 1, 0)));
        }
        if (this.getBlock(pos.add(-1, 1, 0)).getBlock() == Blocks.STICKY_PISTON) {
            if (this.getBlock(pos.add(1,1,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(1, 1, 0));
            }
            else if (this.getBlock(pos.add(1,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(1, 2, 0));
            }
            else if (this.getBlock(pos.add(0,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0,2,0));
            }
            mc.playerController.onPlayerDamageBlock(pos.add(-1, 1, 0), BlockUtil.getRayTraceFacing(pos.add(-1, 1, 0)));
        }
        if (this.getBlock(pos.add(0, 1, 1)).getBlock() == Blocks.STICKY_PISTON) {
            if (this.getBlock(pos.add(0,1,-1)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0, 1, -1));
            }
            else if (this.getBlock(pos.add(0,2,-1)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0, 2, -1));
            }
            else if (this.getBlock(pos.add(0,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0,2,0));
            }
            mc.playerController.onPlayerDamageBlock(pos.add(0, 1, 1), BlockUtil.getRayTraceFacing(pos.add(0, 1, 1)));
        }
        if (this.getBlock(pos.add(0, 1, -1)).getBlock() == Blocks.STICKY_PISTON) {
            if (this.getBlock(pos.add(0,1,1)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0, 1, 1));
            }
            else if (this.getBlock(pos.add(0,2,1)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0, 2, 1));
            }
            else if (this.getBlock(pos.add(0,2,0)).getBlock()== Blocks.AIR){
                this.perform(pos.add(0,2,0));
            }
            mc.playerController.onPlayerDamageBlock(pos.add(0, 1, -1), BlockUtil.getRayTraceFacing(pos.add(0, 1, -1)));
        }
    }

    private void switchToSlot(int slot) {
        if (fullNullCheck()) return;
        AntiPiston.mc.player.connection.sendPacket(new CPacketHeldItemChange(slot));
        AntiPiston.mc.player.inventory.currentItem = slot;
        AntiPiston.mc.playerController.updateController();
    }

    private IBlockState getBlock(BlockPos block) {
        return AntiPiston.mc.world.getBlockState(block);
    }


    private void perform(BlockPos pos) {
        if (fullNullCheck()) return;
        int old = AntiPiston.mc.player.inventory.currentItem;
        this.switchToSlot(this.obsidian);
        BlockUtil.placeBlock(pos, EnumHand.MAIN_HAND, this.rotate.getValue(), true, false);
        this.switchToSlot(old);
    }
}

File: src/main/java/me/dyzjct/kura/module/modules/misc/AntiPiston.kt
package me.dyzjct.kura.module.modules.misc

import me.dyzjct.kura.event.events.block.BlockBreakEvent
import me.dyzjct.kura.module.Category
import me.dyzjct.kura.module.Module
import me.dyzjct.kura.utils.NTMiku.BlockUtil
import me.dyzjct.kura.utils.inventory.InventoryUtil
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.Entity
import net.minecraft.entity.item.EntityEnderCrystal
import net.minecraft.init.Blocks
import net.minecraft.network.play.client.CPacketAnimation
import net.minecraft.network.play.client.CPacketHeldItemChange
import net.minecraft.network.play.client.CPacketUseEntity
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import java.util.stream.Collectors

@Module.Info(name = "AntiPiston", category = Category.MISC)
class AntiPiston : Module() {
    private val rotate = bsetting("Rotate", false)
    private val breakpiston = bsetting("BreakPiston",false)
    private var obsidian = -1
    private var breakpos :BlockPos? = null

    override fun onUpdate() {
        if (fullNullCheck()) return
        if (mc.player == null || mc.world == null) {
            return
        }
        obsidian = InventoryUtil.findHotbarBlock(Blocks.OBSIDIAN)
        if (obsidian == -1) {
            return
        }
        val pos = BlockPos(mc.player.posX, mc.player.posY, mc.player.posZ) ?: return
        if (getBlock(pos.add(1, 1, 0)).block === Blocks.PISTON) {
            if (getBlock(pos.add(-1, 0, 0)).block === Blocks.AIR) {
                perform(pos.add(-1, 0, 0))
            } else if (getBlock(pos.add(-1, 1, 0)).block === Blocks.AIR) {
                perform(pos.add(-1, 1, 0))
            } else if (getBlock(pos.add(-1, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(-1, 2, 0))
            } else if (getBlock(pos.add(0, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(0, 2, 0))
            }
            if (breakpiston.value){
                mc.playerController.onPlayerDamageBlock(pos.add(1, 1, 0), BlockUtil.getRayTraceFacing(pos.add(1, 1, 0)))
            }
        }
        if (getBlock(pos.add(-1, 1, 0)).block === Blocks.PISTON) {
            if (getBlock(pos.add(1, 0, 0)).block === Blocks.AIR) {
                perform(pos.add(1, 0, 0))
            } else if (getBlock(pos.add(1, 1, 0)).block === Blocks.AIR) {
                perform(pos.add(1, 1, 0))
            } else if (getBlock(pos.add(1, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(1, 2, 0))
            } else if (getBlock(pos.add(0, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(0, 2, 0))
            }
            if (breakpiston.value){
                mc.playerController.onPlayerDamageBlock(pos.add(-1, 1, 0), BlockUtil.getRayTraceFacing(pos.add(-1, 1, 0)))
            }
        }
        if (getBlock(pos.add(0, 1, 1)).block === Blocks.PISTON) {
            if (getBlock(pos.add(0, 0, -1)).block === Blocks.AIR) {
                perform(pos.add(0, 0, -1))
            } else if (getBlock(pos.add(0, 1, -1)).block === Blocks.AIR) {
                perform(pos.add(0, 1, -1))
            } else if (getBlock(pos.add(0, 2, -1)).block === Blocks.AIR) {
                perform(pos.add(0, 2, -1))
            } else if (getBlock(pos.add(0, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(0, 2, 0))
            }
            if (breakpiston.value){
                mc.playerController.onPlayerDamageBlock(pos.add(0, 1, 1), BlockUtil.getRayTraceFacing(pos.add(0, 1, 1)))
            }
        }
        if (getBlock(pos.add(0, 1, -1)).block === Blocks.PISTON) {
            if (getBlock(pos.add(0, 0, 1)).block === Blocks.AIR) {
                perform(pos.add(0, 0, 1))
            } else if (getBlock(pos.add(0, 1, 1)).block === Blocks.AIR) {
                perform(pos.add(0, 1, 1))
            } else if (getBlock(pos.add(0, 2, 1)).block === Blocks.AIR) {
                perform(pos.add(0, 2, 1))
            } else if (getBlock(pos.add(0, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(0, 2, 0))
            }
            if (breakpiston.value){
                mc.playerController.onPlayerDamageBlock(pos.add(0, 1, -1), BlockUtil.getRayTraceFacing(pos.add(0, 1, -1)))
            }
        }
        //        STICKY_PISTON
        if (getBlock(pos.add(1, 1, 0)).block === Blocks.STICKY_PISTON) {
            if (getBlock(pos.add(-1, 0, 0)).block === Blocks.AIR) {
                perform(pos.add(-1, 0, 0))
            } else if (getBlock(pos.add(-1, 1, 0)).block === Blocks.AIR) {
                perform(pos.add(-1, 1, 0))
            } else if (getBlock(pos.add(-1, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(-1, 2, 0))
            } else if (getBlock(pos.add(0, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(0, 2, 0))
            }
            if (breakpiston.value){
                mc.playerController.onPlayerDamageBlock(pos.add(1, 1, 0), BlockUtil.getRayTraceFacing(pos.add(1, 1, 0)))
            }
        }
        if (getBlock(pos.add(-1, 1, 0)).block === Blocks.STICKY_PISTON) {
            if (getBlock(pos.add(1, 0, 0)).block === Blocks.AIR) {
                perform(pos.add(1, 0, 0))
            } else if (getBlock(pos.add(1, 1, 0)).block === Blocks.AIR) {
                perform(pos.add(1, 1, 0))
            } else if (getBlock(pos.add(1, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(1, 2, 0))
            } else if (getBlock(pos.add(0, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(0, 2, 0))
            }
            if (breakpiston.value){
                mc.playerController.onPlayerDamageBlock(pos.add(-1, 1, 0), BlockUtil.getRayTraceFacing(pos.add(-1, 1, 0)))
            }
        }
        if (getBlock(pos.add(0, 1, 1)).block === Blocks.STICKY_PISTON) {
            if (getBlock(pos.add(0, 0, -1)).block === Blocks.AIR) {
                perform(pos.add(0, 0, -1))
            } else if (getBlock(pos.add(0, 1, -1)).block === Blocks.AIR) {
                perform(pos.add(0, 1, -1))
            } else if (getBlock(pos.add(0, 2, -1)).block === Blocks.AIR) {
                perform(pos.add(0, 2, -1))
            } else if (getBlock(pos.add(0, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(0, 2, 0))
            }
            if (breakpiston.value){
                mc.playerController.onPlayerDamageBlock(pos.add(0, 1, 1), BlockUtil.getRayTraceFacing(pos.add(0, 1, 1)))
            }
        }
        if (getBlock(pos.add(0, 1, -1)).block === Blocks.STICKY_PISTON) {
            if (getBlock(pos.add(0, 1, 1)).block === Blocks.AIR) {
                perform(pos.add(0, 1, 1))
            } else if (getBlock(pos.add(0, 1, 1)).block === Blocks.AIR) {
                perform(pos.add(0, 1, 1))
            } else if (getBlock(pos.add(0, 2, 1)).block === Blocks.AIR) {
                perform(pos.add(0, 2, 1))
            } else if (getBlock(pos.add(0, 2, 0)).block === Blocks.AIR) {
                perform(pos.add(0, 2, 0))
            }
            if (breakpiston.value){
                mc.playerController.onPlayerDamageBlock(pos.add(0, 1, -1), BlockUtil.getRayTraceFacing(pos.add(0, 1, -1)))
            }
        }
    }

    private fun switchToSlot(slot: Int) {
        if (fullNullCheck()) return
        mc.player.connection.sendPacket(CPacketHeldItemChange(slot))
        mc.player.inventory.currentItem = slot
        mc.playerController.updateController()
    }

    private fun getBlock(block: BlockPos): IBlockState {
        return mc.world.getBlockState(block)
    }

    private fun perform(pos: BlockPos) {
        if (fullNullCheck()) return
        if (pos == breakpos) return
        val old = mc.player.inventory.currentItem
        switchToSlot(obsidian)
        BlockUtil.placeBlock(pos, EnumHand.MAIN_HAND, rotate.value, true, false)
        switchToSlot(old)
    }

    companion object {
        fun breakcrystal() {
            if (fullNullCheck()) return
            for (crystal in mc.world.loadedEntityList.stream()
                .filter { e: Entity -> e is EntityEnderCrystal && !e.isDead }
                .sorted(Comparator.comparing { e: Entity? -> java.lang.Float.valueOf(mc.player.getDistance(e)) })
                .collect(
                    Collectors.toList()
                )) {
                if (crystal !is EntityEnderCrystal || mc.player.getDistance(crystal) > 4.0f) continue
                mc.player.connection.sendPacket(CPacketUseEntity(crystal))
                mc.player.connection.sendPacket(CPacketAnimation(EnumHand.OFF_HAND))
            }
        }
    }


    @SubscribeEvent
    fun onBreak(event: BlockBreakEvent) {
        if (fullNullCheck()) {
            return
        }
        if (event.position != null) {
            breakpos = event.position
        }
    }

}